Flaky-test: ZKSessionTest.testReacquireLeadershipAfterSessionLost

ZKSessionTest is flaky. The testReacquireLeadershipAfterSessionLost test method fails sporadically.

[example failure](https://github.com/apache/pulsar/pull/11026/checks?check_run_id=2889407530#step:9:4861)

```
Error:  Tests run: 6, Failures: 1, Errors: 0, Skipped: 4, Time elapsed: 31.46 s <<< FAILURE! - in org.apache.pulsar.metadata.ZKSessionTest
Error:  testReacquireLeadershipAfterSessionLost(org.apache.pulsar.metadata.ZKSessionTest)  Time elapsed: 15.396 s  <<< FAILURE!
org.awaitility.core.ConditionTimeoutException: Assertion condition defined as a lambda expression in org.apache.pulsar.metadata.ZKSessionTest that uses org.apache.pulsar.metadata.api.coordination.LeaderElection expected [Leading] but found [NoLeader] within 10 seconds.
	at org.awaitility.core.ConditionAwaiter.await(ConditionAwaiter.java:165)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:119)
	at org.awaitility.core.AssertionCondition.await(AssertionCondition.java:31)
	at org.awaitility.core.ConditionFactory.until(ConditionFactory.java:895)
	at org.awaitility.core.ConditionFactory.untilAsserted(ConditionFactory.java:679)
	at org.apache.pulsar.metadata.ZKSessionTest.testReacquireLeadershipAfterSessionLost(ZKSessionTest.java:174)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.AssertionError: expected [Leading] but found [NoLeader]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:617)
	at org.apache.pulsar.metadata.ZKSessionTest.lambda$testReacquireLeadershipAfterSessionLost$1(ZKSessionTest.java:175)
	at org.awaitility.core.AssertionCondition.lambda$new$0(AssertionCondition.java:53)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:222)
	at org.awaitility.core.ConditionAwaiter$ConditionPoller.call(ConditionAwaiter.java:209)
	... 4 more
```

[Another exception in the same test](https://github.com/apache/pulsar/pull/11026/checks?check_run_id=2889407530#step:9:3196)

```
Error:  Tests run: 6, Failures: 1, Errors: 0, Skipped: 4, Time elapsed: 17.285 s <<< FAILURE! - in org.apache.pulsar.metadata.ZKSessionTest
Error:  testReacquireLeadershipAfterSessionLost(org.apache.pulsar.metadata.ZKSessionTest)  Time elapsed: 3.037 s  <<< FAILURE!
java.lang.AssertionError: expected [null] but found [NoLeader]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotSame(Assert.java:1033)
	at org.testng.Assert.assertNull(Assert.java:965)
	at org.testng.Assert.assertNull(Assert.java:953)
	at org.apache.pulsar.metadata.ZKSessionTest.testReacquireLeadershipAfterSessionLost(ZKSessionTest.java:179)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

--- a/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/coordinator/TransactionMetaStoreAssignmentTest.java
+++ b/pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/coordinator/TransactionMetaStoreAssignmentTest.java
@@ -20,6 +20,7 @@ package org.apache.pulsar.broker.transaction.coordinator;
 
 import java.io.IOException;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.List;
 import org.apache.pulsar.broker.PulsarService;
 import org.awaitility.Awaitility;
@@ -32,12 +33,13 @@ public class TransactionMetaStoreAssignmentTest extends TransactionMetaStoreTest
     @Test(groups = "broker")
     public void testTransactionMetaStoreAssignAndFailover() throws IOException {
 
-        int transactionMetaStoreCount = 0;
-        for (PulsarService pulsarService : pulsarServices) {
-            transactionMetaStoreCount += pulsarService.getTransactionMetadataStoreService().getStores().size();
-        }
-
-        Assert.assertEquals(transactionMetaStoreCount, 16);
+        Awaitility.await()
+                .untilAsserted(() -> {
+                    int transactionMetaStoreCount = Arrays.stream(pulsarServices)
+                            .mapToInt(pulsarService -> pulsarService.getTransactionMetadataStoreService().getStores().size())
+                            .sum();
+                    Assert.assertEquals(transactionMetaStoreCount, 16);
+                });
 
         PulsarService crashedMetaStore = null;
         for (int i = pulsarServices.length - 1; i >= 0; i--) {
@@ -62,16 +64,11 @@ public class TransactionMetaStoreAssignmentTest extends TransactionMetaStoreTest
 
         Awaitility.await()
                 .untilAsserted(() -> {
-
-                    int transactionMetaStoreCount2 = 0;
-                    for (PulsarService pulsarService : pulsarServices) {
-                        transactionMetaStoreCount2 += pulsarService.getTransactionMetadataStoreService().getStores()
-                                .size();
-                    }
-
+                    int transactionMetaStoreCount2 = Arrays.stream(pulsarServices)
+                            .mapToInt(pulsarService -> pulsarService.getTransactionMetadataStoreService().getStores().size())
+                            .sum();
                     Assert.assertEquals(transactionMetaStoreCount2, 16);
                 });
-
         transactionCoordinatorClient.close();
     }
 }

--- a/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
+++ b/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
@@ -178,7 +178,6 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
         Thread.sleep(2_000);
         Awaitility.await()
                 .untilAsserted(()-> assertEquals(le1.getState(),LeaderElectionState.Leading));
-        assertEquals(le1.getState(),LeaderElectionState.Leading);
         assertTrue(store.get(path).join().isPresent());
     }
 }

--- a/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
+++ b/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
@@ -136,6 +136,7 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
 
     @Test
     public void testReacquireLeadershipAfterSessionLost() throws Exception {
+        //  ---  init
         @Cleanup
         MetadataStoreExtended store = MetadataStoreExtended.create(zks.getConnectionString(),
                 MetadataStoreConfig.builder()
@@ -144,7 +145,6 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
 
         BlockingQueue<SessionEvent> sessionEvents = new LinkedBlockingQueue<>();
         store.registerSessionListener(sessionEvents::add);
-
         BlockingQueue<LeaderElectionState> leaderElectionEvents = new LinkedBlockingQueue<>();
         String path = newKey();
 
@@ -153,13 +153,13 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
         @Cleanup
         LeaderElection<String> le1 = coordinationService.getLeaderElection(String.class, path,
                 leaderElectionEvents::add);
-
+        // --- test manual elect
         le1.elect("value-1").join();
         assertEquals(le1.getState(), LeaderElectionState.Leading);
 
         LeaderElectionState les = leaderElectionEvents.poll(5, TimeUnit.SECONDS);
         assertEquals(les, LeaderElectionState.Leading);
-
+        // --- expire session
         zks.expireSession(((ZKMetadataStore) store).getZkSessionId());
 
         SessionEvent e = sessionEvents.poll(5, TimeUnit.SECONDS);
@@ -167,28 +167,18 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
 
         e = sessionEvents.poll(10, TimeUnit.SECONDS);
         assertEquals(e, SessionEvent.SessionLost);
-
-        // due to zookeeper async notice, we need to wait.
+        // --- test  le1 can be leader
         Awaitility.await()
-                .until(() -> le1.getState() == LeaderElectionState.Leading);
-        assertEquals(le1.getState(), LeaderElectionState.Leading);
-        Awaitility.await()
-                .until(()-> leaderElectionEvents.poll() == null);
-        les = leaderElectionEvents.poll();
-        assertNull(les);
-
+                .untilAsserted(()-> assertEquals(le1.getState(),LeaderElectionState.Leading)); // reacquire leadership
         e = sessionEvents.poll(10, TimeUnit.SECONDS);
         assertEquals(e, SessionEvent.Reconnected);
         e = sessionEvents.poll(10, TimeUnit.SECONDS);
         assertEquals(e, SessionEvent.SessionReestablished);
-
-        // due to zookeeper async notice, we need to wait.
+        // ---- after 2000 millis, the le1 is also can be leader.
+        Thread.sleep(2_000);
         Awaitility.await()
-                .until(() -> le1.getState() == LeaderElectionState.Leading);
-        assertEquals(le1.getState(), LeaderElectionState.Leading);
-        les = leaderElectionEvents.poll();
-        assertNull(les);
-
+                .untilAsserted(()-> assertEquals(le1.getState(),LeaderElectionState.Leading));
+        assertEquals(le1.getState(),LeaderElectionState.Leading);
         assertTrue(store.get(path).join().isPresent());
     }
 }

--- a/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
+++ b/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
@@ -174,8 +174,6 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
         assertEquals(e, SessionEvent.Reconnected);
         e = sessionEvents.poll(10, TimeUnit.SECONDS);
         assertEquals(e, SessionEvent.SessionReestablished);
-        // ---- after 2000 millis, the le1 is also can be leader.
-        Thread.sleep(2_000);
         Awaitility.await()
                 .untilAsserted(()-> assertEquals(le1.getState(),LeaderElectionState.Leading));
         assertTrue(store.get(path).join().isPresent());

--- a/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
+++ b/pulsar-metadata/src/test/java/org/apache/pulsar/metadata/ZKSessionTest.java
@@ -22,13 +22,10 @@ import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertNull;
 import static org.testng.Assert.assertTrue;
-
 import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.TimeUnit;
-
 import lombok.Cleanup;
-
 import org.apache.pulsar.metadata.api.MetadataStoreConfig;
 import org.apache.pulsar.metadata.api.coordination.CoordinationService;
 import org.apache.pulsar.metadata.api.coordination.LeaderElection;
@@ -171,10 +168,12 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
         e = sessionEvents.poll(10, TimeUnit.SECONDS);
         assertEquals(e, SessionEvent.SessionLost);
 
-        Awaitility.await().untilAsserted(() -> {
-            assertEquals(le1.getState(), LeaderElectionState.Leading);
-        });
-
+        // due to zookeeper async notice, we need to wait.
+        Awaitility.await()
+                .until(() -> le1.getState() == LeaderElectionState.Leading);
+        assertEquals(le1.getState(), LeaderElectionState.Leading);
+        Awaitility.await()
+                .until(()-> leaderElectionEvents.poll() == null);
         les = leaderElectionEvents.poll();
         assertNull(les);
 
@@ -183,9 +182,10 @@ public class ZKSessionTest extends BaseMetadataStoreTest {
         e = sessionEvents.poll(10, TimeUnit.SECONDS);
         assertEquals(e, SessionEvent.SessionReestablished);
 
-        Awaitility.await().untilAsserted(() -> {
-                    assertEquals(le1.getState(), LeaderElectionState.Leading);
-        });
+        // due to zookeeper async notice, we need to wait.
+        Awaitility.await()
+                .until(() -> le1.getState() == LeaderElectionState.Leading);
+        assertEquals(le1.getState(), LeaderElectionState.Leading);
         les = leaderElectionEvents.poll();
         assertNull(les);
 

